app/drivers/kscan/kscan_gpio_matrix.c

```diff
@@ -51,257 +158,300 @@ static int kscan_gpio_config_interrupts(const struct device **devices,
+            const int index = state_index_rc(config, r, c);
+            const bool pressed = data->next_state[index];
+
+            // Follow up reads are needed if any key is pressed because further
+            // interrupts won't fire on already tripped GPIO pins.
+#if USE_INTERRUPTS
+            submit_followup_read = submit_followup_read || pressed;
+#endif
+            if (pressed != data->current_state[index]) {
+                LOG_DBG("Sending event at %i,%i state %s", r, c, pressed ? "on" : "off");
+                data->current_state[index] = pressed;
+                data->callback(dev, r, c, pressed);
+            }
+        }
+    }
+
+#if USE_INTERRUPTS
+    if (submit_followup_read) {
+        // At least one key is pressed. Poll until everything is released.
+        // TODO (Zephyr 2.6): use k_work_reschedule()
+        k_delayed_work_cancel(&data->work);
+        k_delayed_work_submit(&data->work, K_MSEC(config->debounce_period_ms));
+    } else {
+        // All keys are released. Return to waiting for an interrupt.
+        kscan_matrix_interrupt_enable(dev);
+    }
+#endif
+
+    return 0;
+}
+
+#if USE_POLLING
+static void kscan_matrix_timer_handler(struct k_timer *timer) {
+    struct kscan_matrix_data *data = CONTAINER_OF(timer, struct kscan_matrix_data, poll_timer);
+    k_delayed_work_submit(&data->work, K_NO_WAIT);
+}
+#endif
+
+static void kscan_matrix_work_handler(struct k_work *work) {
+    struct k_delayed_work *dwork = CONTAINER_OF(work, struct k_delayed_work, work);
+    struct kscan_matrix_data *data = CONTAINER_OF(dwork, struct kscan_matrix_data, work);
+    kscan_matrix_read(data->dev);
+}
+
+static int kscan_matrix_configure(const struct device *dev, const kscan_callback_t callback) {
+    struct kscan_matrix_data *data = dev->data;
+
+    if (!callback) {
+        return -EINVAL;
+    }
+
+    data->callback = callback;
+    return 0;
+}
+
+static int kscan_matrix_enable(const struct device *dev) {
+#if USE_POLLING
+    struct kscan_matrix_data *data = dev->data;
+    const struct kscan_matrix_config *config = dev->config;
+
+    k_timer_start(&data->poll_timer, K_MSEC(config->poll_period_ms),
+                  K_MSEC(config->poll_period_ms));
+    return 0;
+#else
+    // Read will automatically enable interrupts once done.
+    return kscan_matrix_read(dev);
+#endif
+}
+
+static int kscan_matrix_disable(const struct device *dev) {
+#if USE_POLLING
+    struct kscan_matrix_data *data = dev->data;
+
+    k_timer_stop(&data->poll_timer);
+    return 0;
+#else
+    return kscan_matrix_interrupt_disable(dev);
+#endif
+}
+
+static int kscan_matrix_init_input_inst(const struct device *dev,
+                                        const struct kscan_gpio_dt_spec *gpio, const int index) {
+    if (!device_is_ready(gpio->port)) {
+        LOG_ERR("GPIO is not ready: %s", gpio->port->name);
+        return -ENODEV;
+    }
+
+    int err = gpio_pin_configure(gpio->port, gpio->pin, GPIO_INPUT | gpio->dt_flags);
```

Comment:
What are your thoughts on inferring all of the GPIO flags for inputs (and outputs)? Instead of manually defining them in our DT, we could infer inputs to have `GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN` and outputs to have `GPIO_ACTIVE_HIGH`. We know this based on the `diode-direction` property. To me it seems that `diode-direction` and these GPIO flags explain the same thing, so we should only need one. This would result in a cleaner keyboard DT.

Reply:
Yeah, I was considering adding that here, but figured it would be better to do that as a separate commit. I think we'd want to infer the flags only if `gpio->dt_flags` is 0 so you can still use active high, pull up if needed.

Reply:
I would rather we do this as a follow up. We've discussed this previously, and I'm *mostly* open to it, but we definitely have one scenario where that is *flipped*, which is Ferris, where the IO expander only has a pull down, so the diode direct does *not* actually infer those flags. Not saying we don't do it, just needs to be clearly documented if we do.

Reply:
I was going to suggest that we do `GPIO_ACTIVE_LOW` + `GPIO_ACTIVE_LOW | GPIO_PULL_UP` by default since I knew that atmegas only have pull-ups, but since we don't support atmegas, I didn't. I guess this still applies to things like these IO expanders and other older/simpler ICs. Generally, it seems that pull-ups are found everywhere though, so it might be a saner default.

Reply:
After looking at this more, I think we'd need to create our own "auto" define. `GPIO_ACTIVE_HIGH` is 0, so using 0 would make it impossible to select just `GPIO_ACTIVE_HIGH` (e.g. when you have an external pull down) instead of `GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN`.

Reply:
Since it's a bit field... we would risk trying to use an unused bit here to flag a "auto". Let's discuss further on a follow up.